---
title: "Modèles mixtes"
subtitle: "TP1: `GSE72680`"
author: "Florent Chuffart"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
---


```{r, echo=FALSE, eval=TRUE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=TRUE, results="verbatim", warning=FALSE)
```
# Data

J’étudie le jeu de données suivant : 

```{r data, eval=TRUE, echo=TRUE, results="verbatim"}
d = datasets::Orange
d$Tree = factor(d$Tree, levels=unique(d$Tree), order=FALSE)
# head(d)
# dim(d)
d2 = d

gse = "GSE72680"
print(paste0("clinical data from epimeddb: ", gse))
url = paste0("http://epimed.univ-grenoble-alpes.fr/database/parameters/",gse)
# df1 = read.csv2(url, header=TRUE, sep=";", stringsAsFactors=FALSE, dec=".", na.strings="", row.names=1)
# saveRDS(df1, "GSE72680_df1.rds")
df1 = readRDS("GSE72680_df1.rds")
# url = paste0("http://epimed.univ-grenoble-alpes.fr/database/expgroup/",gse)
# df2 = read.csv2(url, header=TRUE, sep=";", stringsAsFactors=FALSE, dec=".",na.strings="", row.names=1)

d = df1

# remove NA...
#   ... for rows
dim(d)
# d = d[!apply(is.na(d[,c(
#   "text",
#   # "childhood_sexual_physical_abuse_moderate_to_extreme",
#   # "treatment_for_depression"                           ,
#   # "treatment_for_bipolar_disorder"                     ,
#   # "treatment_for_posttraumatic_stress_disorder"        ,
#   # "treatment_for_anxiety_disorder"                     ,
#   "body_mass_index"
#   )]), 1, any),]
# dim(d)
# d = d[!apply(apply(d[,c(
#   "text",
#   # "childhood_sexual_physical_abuse_moderate_to_extreme",
#   # "treatment_for_depression"                           ,
#   # "treatment_for_bipolar_disorder"                     ,
#   # "treatment_for_posttraumatic_stress_disorder"        ,
#   # "treatment_for_anxiety_disorder"                     ,
#   "body_mass_index"
#   )], 1, "%in%", "--"), 2, any),]
# d$body_mass_index = as.numeric(d$body_mass_index)
dim(d)
#   ... for columns
dim(d)
layout(matrix(1:2, 1), respect=TRUE)
foo = apply(is.na(d), 2, sum)
table(foo)
barplot(table(foo), las=2)
# head(d)
dim(d)
d = d[,!apply(is.na(d), 2, any)]
foo = apply(is.na(d), 2, sum)
table(foo)
barplot(table(foo), las=2)
# head(d)
dim(d)

d = d[!d$race_ethnicity %in% c("Mixed", "Other"),]



colnames(d)[colnames(d)%in%"dna_methylation_predicted_age"] = "dnamage"
# d$text
d$sex = factor(d$sex, levels=unique(d$sex), order=FALSE)
# d$age
# d$dnamage
# d$cd8_t_cells
# d$cd4_t_cells
# d$natural_killer_cells
# d$b_cells
# d$monocytes
# d$granulocytes
# d$tissue
d$race_ethnicity     =  factor(d$race_ethnicity, levels=unique(d$race_ethnicity), order=FALSE)

table(d$sex        )
table(d$race_ethnicity)
table(d$sex, d$race_ethnicity)

d$grp = d$sex
# d$grp = d$race_ethnicity
d$grp = paste0(d$sex, "_", d$race_ethnicity)
d$grp = factor(d$grp, levels=unique(d$grp), order=FALSE)

# class
class(d) = class(d2)
attributes(d)$formula = formula(dnamage ~ age | grp)
                                             


layout(1, respect=TRUE)
plot(dnamage~age, d, pch=16, col=d$grp)
head(d[,-1])
table(d$grp)
```

*STAT DESC*

1. Mettre les titres et les légendes des graphiques.






# Modèles

## Modèles lineaires classiques

Je construis les modèles suivants : 

```{r}
m0 = lm(dnamage~age, d)
m1 = lm(dnamage~age+grp, d)
m2 = lm(dnamage~age*grp, d)
```



1. J’écris la formule analytique des modèles proposés ;
2. Je visualise les droites de régression et le graphique des prédictions de chacun des modèles ;
3. J’étudie la distribution des résidus ; 
4. Je valide les hypothèses des modèles ;



## `m0`
```{r echo=FALSE, results="verbatim"}
m0 = lm(dnamage~age, d)
summary(m0)
layout(matrix(1:2, 1, byrow=TRUE), respect=TRUE)
plot(dnamage~age, d, pch=16, col=d$grp)
abline(m0)
```

## `m1`
```{r echo=FALSE, results="verbatim"}
m1 = lm(dnamage~age+grp, d)
# m1 = lm(dnamage~age+sex+race_ethnicity, d)
summary(m1)
layout(matrix(1:2, 1, byrow=TRUE), respect=TRUE)
plot(dnamage~age, d, pch=16, col=d$grp)
abline(a=m1$coef[1]             , b=m1$coef[2], col=1)
n = length(levels(d$grp))
for (i in 2:n) {
  abline(a=m1$coef[1] + m1$coef[i+1], b=m1$coef[2], col=i)  
}
legend("bottomright", cex=.7, levels(d$grp), col=1:length(levels(d$grp)), pch=16)
```

## `m2`
```{r echo=FALSE, results="verbatim"}
m2 = lm(dnamage~age*grp, d)
summary(m2)
layout(matrix(1:2, 1, byrow=TRUE), respect=TRUE)
plot(dnamage~age, d, pch=16, col=d$grp)
abline(a=m2$coef[1]             , b=m2$coef[2]              , col=1)
n = length(levels(d$grp))
for (i in 2:n) {
  abline(a=m2$coef[1] + m2$coef[i+1], b=m2$coef[2] + m2$coef[n+i] , col=i)
}
legend("bottomright", cex=.7, levels(d$grp), col=1:length(levels(d$grp)), pch=16)
```


## Modèles mixtes

Je construis les modèles suivants : 

- effet aléatoire sur l’ordonnée à l’origine et sur la pente
- effet aléatoire uniquement sur l’ordonnée à l’origine 
- effet aléatoire uniquement sur la pente


```{r}
m3 = nlme::lme(dnamage ~ age, data=d)
m4 = lme4::lmer(dnamage ~ age + (age | grp), data=d)
m5 = nlme::lme( dnamage ~ age, random = ~ 1, data = d)
m6 = lme4::lmer(dnamage ~ age + (1 | grp), data = d)
m7 = nlme::lme( dnamage ~ age, random = ~ age-1, data = d)
m8 = lme4::lmer(dnamage ~ age + (0+age | grp), data = d)
```

1. J’écris la formule analytique des modèles proposés ;
2. Je visualise les droites de régression et le graphique des prédictions de chacun des modèles ;
3. J’étudie la distribution des résidus ; 
4. Je valide les hypothèses des modèles ;



```{r lme, echo=FALSE, results="verbatim"}
layout(matrix(1:2, 1, byrow=TRUE), respect=TRUE)
# library(nlme)
# res.lme = lme(distance ~ age, data = Orthodont) summary(res.lme)
m3 = nlme::lme(dnamage ~ age, data=d)
summary(m3)
coef(m3)
plot(dnamage~age, d, pch=16, col=d$grp)
n = length(levels(d$grp))
for (i in 1:n) {
  abline(a=coef(m3)$grp[i,1], b=coef(m3)$grp[i,2], col=i)
}
legend("bottomright", cex=.7, levels(d$grp), col=1:length(levels(d$grp)), pch=16)

m4 = lme4::lmer(dnamage ~ age + (age | grp), data=d)
summary(m4)
coef(m4)
plot(dnamage~age, d, pch=16, col=d$grp)
n = length(levels(d$grp))
for (i in 1:n) {
  abline(a=coef(m4)$grp[i,1], b=coef(m4)$grp[i,2], col=i)
}
legend("bottomright", cex=.7, levels(d$grp), col=1:length(levels(d$grp)), pch=16)

m5 = nlme::lme( dnamage ~ age, random = ~ 1, data = d)
summary(m5)
coef(m5)
plot(dnamage~age, d, pch=16, col=d$grp)
n = length(levels(d$grp))
for (i in 1:n) {
  abline(a=coef(m5)$grp[i,1], b=coef(m5)$grp[i,2], col=i)
}
legend("bottomright", cex=.7, levels(d$grp), col=1:length(levels(d$grp)), pch=16)


m6 = lme4::lmer(dnamage ~ age + (1 | grp), data = d)
summary(m6)
coef(m6)
plot(dnamage~age, d, pch=16, col=d$grp)
n = length(levels(d$grp))
for (i in 1:n) {
  abline(a=coef(m6)$grp[i,1], b=coef(m6)$grp[i,2], col=i)
}
legend("bottomright", cex=.7, levels(d$grp), col=1:length(levels(d$grp)), pch=16)


m7 = nlme::lme( dnamage ~ age, random = ~ age-1, data = d)
summary(m7)
coef(m7)
plot(dnamage~age, d, pch=16, col=d$grp)
n = length(levels(d$grp))
for (i in 1:n) {
  abline(a=coef(m7)$grp[i,1], b=coef(m7)$grp[i,2], col=i)
}
legend("bottomright", cex=.7, levels(d$grp), col=1:length(levels(d$grp)), pch=16)

m8 = lme4::lmer(dnamage ~ age + (0+age | grp), data = d)
summary(m8)
coef(m8)
plot(dnamage~age, d, pch=16, col=d$grp)
n = length(levels(d$grp))
for (i in 1:n) {
  abline(a=coef(m8)$grp[i,1], b=coef(m8)$grp[i,2], col=i)
}
legend("bottomright", cex=.7, levels(d$grp), col=1:length(levels(d$grp)), pch=16)
```



## Choix de modèles


```{r}
BIC(m0)
BIC(m1)
BIC(m2)
BIC(m3)
BIC(m4)
BIC(m5)
BIC(m6)
BIC(m7)
BIC(m8)

plot(0:8, c(
  BICm0=BIC(m0),
  BICm1=BIC(m1),
  BICm2=BIC(m2),
  BICm3=BIC(m3),
  BICm4=BIC(m4),
  BICm5=BIC(m5),
  BICm6=BIC(m6),
  BICm7=BIC(m7),
  BICm8=BIC(m8)
))

```


# Conclusion

1. Conclure


# References


Zannas AS, Jia M, Hafner K, Baumert J et al. *Epigenetic upregulation of FKBP5 by aging and stress contributes to NF-κB-driven inflammation and cardiovascular risk.* **PNAS** U S A 2019 Jun 4;116(23):11370-11379. PMID: 31113877

**GSE72680** K. Ressler, A. Smith, E. Binder, V. Kilaru, *DNA Methylation of African Americans from the Grady Trauma Project.* Gene Expression Omnibus (GEO). https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE72680. Deposited 3 September 2015.


# Session Information

```{r results="verbatim"}
sessionInfo()
```

